### PR TITLE
Topic recognition - Charles Gore s4436638

### PR DESCRIPTION
**Pull request by CharlieGore**
Pull request for the Brain MRI Super Resolution Network on the ADNI dataset.

This is a test to demonstrate the pull request and its marking during course COMP3710 at UQ.

This contains:
-A dataset script for handling the dataset.
-A training script for training the neural network.
-A predict script to use the neural network on a fresh dataset to test effectiveness.
-A modules script containing functions for the other scripts.

The following requirements by the maintainers have been met:

-topic-recognition branch
-Commit log history
-Rest of the marking criteria

Requires PyTorch 1.13 or above

Regards,
Charles Gore (s4436638)

### CODE REVIEW

**Comment by uqmarlonbran** [comment]
**This is an initial inspection - no action is required yet**

**Difficulty:** Normal

**General Comments:** Nice approach overall, creates a custom dataloader to deal with AD and NC file structure, as well as allow for validation subset. Code exists to compute and store metrics on the test set, though they are not reported in the README. 

**[Recognition Problem]**
"modules.py":

- Implements a multi-stage upscaling process, achieved with dedicated upscale blocks and convolution blocks.
- Good use of sequential blocks.
- Adds a final convolution layer after the last upscaling that reduces feature size (rather than pixelshuffle reducing down to 1).

"dataset.py":

- Creates custom datasets to deal with AD and NC file structure.
- Allows for training and validation sets.

"train.py":

- Keeps track of training and validation losses, providing a plot when training completed.
- Saves the model with the lowest validation loss.
- Also stores train and val losses to csv file.

"predict.py":

- Computes PSNR, SSIM and MSE to all images in test set.
- Saves images to dedicated folder upon completion.
- Stores metrics to csv.

"README.MD":

- Provides a brief overview of PixelCNN and its implementation.
- Provides some motivation to 2x -> 2x upscale, though no ablation demonstrated.
- Provides file structure.
- Provides dependencies.
- Provides usage example.
- Output image seems in-line with other submissions. 

**[Commit Log]**
Commit Log looks ok.

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Normal -5
## Commit Log
Meaningful commit messages sometimes -1
Progressive commits used
## Documentation
ReadMe acceptable, no refs -1
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
